### ISSUE TITLE
TypeError: Axis must be specified when shapes of a and weights differ  - Differentially Private EBMs

### ISSUE THREAD

**Issue by mtaufeeq** [interpretml/interpret#349]
Hi, 

With this lines of code

`dp_ebm = DPExplainableBoostingClassifier(epsilon=1, delta=1e-5) # Specify privacy parameters
dp_ebm.fit(train_df[features], train_df[target])`

I am getting this Type error. Any ideas on how to fix this? 

`---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-13-19fbfb90a84a> in <module>
      1 dp_ebm = DPExplainableBoostingClassifier(epsilon=1, delta=1e-5) # Specify privacy parameters
----> 2 dp_ebm.fit(train_df[features], train_df[target])
      3 
      4 # show(dp_ebm.explain_global()) # Identical function calls to standard EBMs

~\.conda\envs\rdevs\lib\site-packages\interpret\glassbox\ebm\ebm.py in fit(self, X, y, sample_weight)
   1196             # DP method of generating feature importances can generalize to non-dp if preprocessors start tracking joint distributions
   1197             for i in range(len(self.feature_groups_)):
-> 1198                 mean_abs_score = np.average(np.abs(self.additive_terms_[i]), weights=self.preprocessor_.col_bin_counts_[i])
   1199                 self.feature_importances_.append(mean_abs_score)
   1200         else:

~\.conda\envs\rdevs\lib\site-packages\numpy\core\overrides.py in average(*args, **kwargs)

~\.conda\envs\rdevs\lib\site-packages\numpy\lib\function_base.py in average(a, axis, weights, returned)
    506         if a.shape != wgt.shape:
    507             if axis is None:
--> 508                 raise TypeError(
    509                     "Axis must be specified when shapes of a and weights "
    510                     "differ.")

TypeError: Axis must be specified when shapes of a and weights differ.`

**Comment by interpret-ml** [comment]
Hi @mtaufeeq, 

I haven't seen this error before, and am having trouble reproducing it with any public dataset. For example, the following code runs for me on the currently available PyPI release of interpret (v0.2.7):

```python3
import pandas as pd
from sklearn.model_selection import train_test_split

df = pd.read_csv(
    "https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data",
    header=None)
df.columns = [
    "Age", "WorkClass", "fnlwgt", "Education", "EducationNum",
    "MaritalStatus", "Occupation", "Relationship", "Race", "Gender",
    "CapitalGain", "CapitalLoss", "HoursPerWeek", "NativeCountry", "Income"
]

train_cols = df.columns[0:-1]
label = df.columns[-1]
X = df[train_cols]
y = df[label].apply(lambda x: 0 if x == " <=50K" else 1) #Turning response into 0 and 1

seed = 1
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=seed)

from interpret.privacy import DPExplainableBoostingClassifier
from interpret.glassbox import ExplainableBoostingClassifier
import time
from sklearn.metrics import roc_auc_score, accuracy_score

start = time.time()
dpebm = DPExplainableBoostingClassifier(epsilon=1, delta=1e-6)
_ = dpebm.fit(X_train, y_train)

dp_auroc = roc_auc_score(y_test, dpebm.predict_proba(X_test)[:, 1])
end = time.time()

print(f"DP EBM with eps: {dpebm.epsilon} and delta: {dpebm.delta} trained in {end - start:.2f} seconds with a test AUC of {dp_auroc:.3f}")


start = time.time()
ebm = ExplainableBoostingClassifier()
_ = ebm.fit(X_train, y_train)

ebm_auroc = roc_auc_score(y_test, ebm.predict_proba(X_test)[:, 1])
end = time.time()
print(f"EBM trained in {end - start:.2f} seconds with a test AUC of {ebm_auroc:.3f}")
```

The full example I ran this from can be found in this notebook: https://github.com/interpretml/interpret/blob/develop/examples/python/notebooks/Differentially%20Private%20EBMs.ipynb


Are you able to share any details of your dataset (e.g. number of samples, number of features, how many numeric vs. categorical)? Also, did you install interpretML from PyPI (`pip install interpret`) or build a pre-release version from source?

As a sanity check, it might be worth verifying that a standard EBM (without differential privacy) or another scikit-learn model can train on the exact same data without receiving an error, e.g.:

```python3
from interpret.glassbox import ExplainableBoostingClassifier

ebm = ExplainableBoostingClassifier()
ebm.fit(train_df[features], train_df[target])
```
Appreciate you reporting this!

**Comment by interpret-ml** [comment]
Hi @mtaufeeq,

Is your classification problem multiclass?  Currently we only support binary classification and regression in the DP-EBM setting.  Our last PyPI release would probably raise this rather confusing exception given multiclass inputs.  The latest code in the develop branch raises a much better and more descriptive exception.

-InterpretML Team
